common/src/main/cpp/extensions/openxr_fb_hand_tracking_mesh_extension_wrapper.cpp

```diff
@@ -0,0 +1,325 @@
+	if (singleton == nullptr) {
+		singleton = memnew(OpenXRFbHandTrackingMeshExtensionWrapper());
+	}
+	return singleton;
+}
+
+OpenXRFbHandTrackingMeshExtensionWrapper::OpenXRFbHandTrackingMeshExtensionWrapper() :
+		OpenXRExtensionWrapperExtension() {
+	ERR_FAIL_COND_MSG(singleton != nullptr, "An OpenXRFbHandTrackingMeshExtensionWrapper singleton already exists.");
+
+	request_extensions[XR_FB_HAND_TRACKING_MESH_EXTENSION_NAME] = &fb_hand_tracking_mesh_ext;
+	singleton = this;
+}
+
+OpenXRFbHandTrackingMeshExtensionWrapper::~OpenXRFbHandTrackingMeshExtensionWrapper() {
+	cleanup();
+}
+
+void OpenXRFbHandTrackingMeshExtensionWrapper::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("is_enabled"), &OpenXRFbHandTrackingMeshExtensionWrapper::is_enabled);
+
+	ADD_SIGNAL(MethodInfo("openxr_fb_hand_tracking_mesh_data_fetched", PropertyInfo(Variant::INT, "hand_index")));
+}
+
+void OpenXRFbHandTrackingMeshExtensionWrapper::cleanup() {
+	fb_hand_tracking_mesh_ext = false;
+
+	for (int i = 0; i < Hand::HAND_MAX; i++) {
+		is_hand_mesh_data_fetched[i] = false;
+		hand_tracking_scale[i].overrideHandScale = false;
+		hand_tracking_scale[i].overrideValueInput = 1.0;
+	}
+}
+
+godot::Dictionary OpenXRFbHandTrackingMeshExtensionWrapper::_get_requested_extensions() {
+	godot::Dictionary result;
+	for (auto ext : request_extensions) {
+		godot::String key = ext.first;
+		uint64_t value = reinterpret_cast<uint64_t>(ext.second);
+		result[key] = (godot::Variant)value;
+	}
+	return result;
+}
+
+void OpenXRFbHandTrackingMeshExtensionWrapper::_on_instance_created(uint64_t instance) {
+	if (fb_hand_tracking_mesh_ext) {
+		bool result = initialize_fb_hand_tracking_mesh_extension((XrInstance)instance);
+		if (!result) {
+			UtilityFunctions::print("Failed to initialize fb_hand_tracking_mesh extension");
+			fb_hand_tracking_mesh_ext = false;
+		}
+	}
+}
+
+void OpenXRFbHandTrackingMeshExtensionWrapper::_on_instance_destroyed() {
+	cleanup();
+}
+
+uint64_t OpenXRFbHandTrackingMeshExtensionWrapper::_set_hand_joint_locations_and_get_next_pointer(int32_t p_hand_index, void *p_next_pointer) {
+	if (!fb_hand_tracking_mesh_ext) {
+		return reinterpret_cast<uint64_t>(p_next_pointer);
+	}
+
+	hand_tracking_scale[p_hand_index].type = XR_TYPE_HAND_TRACKING_SCALE_FB;
+	hand_tracking_scale[p_hand_index].next = p_next_pointer;
+
+	return reinterpret_cast<uint64_t>(&hand_tracking_scale[p_hand_index]);
+}
+
+void OpenXRFbHandTrackingMeshExtensionWrapper::_on_process() {
+	if (!is_enabled() || get_openxr_api().is_null()) {
+		return;
+	}
+
+	for (int i = 0; i < Hand::HAND_MAX; i++) {
+		if (!is_hand_mesh_data_fetched[i]) {
+			XrHandTrackerEXT hand_tracker = reinterpret_cast<XrHandTrackerEXT>(get_openxr_api()->get_hand_tracker(i));
+			if (hand_tracker != XR_NULL_HANDLE) {
+				if (fetch_hand_mesh_data(Hand(i))) {
+					is_hand_mesh_data_fetched[i] = true;
+					emit_signal("openxr_fb_hand_tracking_mesh_data_fetched", i);
+				}
+			}
+		}
+	}
```

Comment:
Does this need to happen in `_on_process()`? Or, could we do it in `_on_state_ready()`? Or, maybe even only after `OpenXRFbHandTrackingMesh` has called a method asking for the mesh (or otherwise indicated that it's interested)?

Not all apps using 'godot_openxr_vendors' will want to use this extension even if it's available, so it'd be best to only do this on demand. Or, if that's not possible, then we should have a project setting to enable or disable it (but those are hard for developers to notice, so that option is not ideal).

Reply:
We discussed this a bit further and this will remain in `on_process()`, but I have changed it so it will only fetch the mesh data if `OpenXRFbHandTrackingMesh` nodes are used.